parachain/pallets/ethereum-beacon-client/src/lib.rs

```diff
@@ -0,0 +1,679 @@
+
+#[derive(Clone, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct PublicKey([u8; 48]);
+
+impl Default for PublicKey {
+	fn default() -> Self {
+		PublicKey([0u8; 48])
+	}
+}
+
+/// Beacon block header as it is stored in the runtime storage. The block root is the
+/// Merklization of a BeaconHeader.
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct BeaconBlockHeader {
+	// The slot for which this block is created. Must be greater than the slot of the block defined by parentRoot.
+	pub slot: u64,
+	// The index of the validator that proposed the block.
+	pub proposer_index: ValidatorIndex,
+	// The block root of the parent block, forming a block chain.
+	pub parent_root: Root,
+	// The hash root of the post state of running the state transition through this block.
+	pub state_root: Root,
+	// The hash root of the beacon block body
+	pub body_root: Root,
+}
+
+/// Sync committee as it is stored in the runtime storage.
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct SyncCommittee {
+	pub pubkeys: Vec<PublicKey>,
+	pub aggregate_pubkey: PublicKey,
+}
+
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct SyncAggregate {
+	// 1 or 0 bit, indicates whether a sync committee participated in a vote
+	pub sync_committee_bits: Vec<u8>,
+	pub sync_committee_signature: Vec<u8>,
+}
+
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct InitialSync {
+	pub header: BeaconBlockHeader,
+	pub current_sync_committee: SyncCommittee,
+	pub current_sync_committee_branch: ProofBranch,
+	pub validators_root: Root,
+}
+
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct SyncCommitteePeriodUpdate {
+	pub attested_header: BeaconBlockHeader,
+	pub next_sync_committee: SyncCommittee,
+	pub next_sync_committee_branch: ProofBranch,
+	pub finalized_header: BeaconBlockHeader,
+	pub finality_branch: ProofBranch,
+	pub sync_aggregate: SyncAggregate,
+	pub fork_version: ForkVersion,
+	pub sync_committee_period: u64,
+}
+
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct FinalizedHeaderUpdate {
+	pub attested_header: BeaconBlockHeader,
+	pub finalized_header: BeaconBlockHeader,
+	pub finality_branch: ProofBranch,
+	pub sync_aggregate: SyncAggregate,
+	pub fork_version: ForkVersion,
+}
+
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct ForkData {
+	// 1 or 0 bit, indicates whether a sync committee participated in a vote
+	pub current_version: [u8; 4],
+	pub genesis_validators_root: [u8; 32],
+}
+
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct SigningData {
+	pub object_root: Root,
+	pub domain: Domain,
+}
+
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct Genesis {
+	pub validators_root: Root,
```

Comment:
Is it possible to store the root directly without the redundant Genesis wrapper? As Genesis could be confused with GenesisConfig

Reply:
Good point, will do that.